### PR TITLE
Resolves #397: Add AutoContinuingCursor to chain cursors across transactions

### LINKED ISSUE #397: Add AutoContinuingCursor to chain cursors across transactions
As @scgray suggested in 
https://github.com/FoundationDB/fdb-record-layer/pull/384#discussion_r257032955, it would be pretty useful to have a cursor to chain cursors across transactions.

### PR DESCRIPTION
**Pull request by nblintao**
- Implemented AutoContinuingCursor and AutoContinuingCursorTest
- Added a helper method noNextButHasContinuation in RecordCursorResult
- Fix doc in ChainedCursor

### CODE REVIEW

**Comment by nblintao** [comment]
Also cc @nschiefer who might be interested in this.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java [hunk 1]
```diff
@@ -200,6 +200,14 @@ public int hashCode() {
         }
     }
 
+    /**
+     * Return whether or not this result does not include a next value but include a continuation (i.e. source is not exhausted).
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java]: This sentence is a little hard to parse.  How about something like:

    Returns true if the cursor has reached its end but a continuation is still available (i.e. the source is not yet exhausted)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+
+/**
+ * A wrapper for a cursor to iterate across multiple transactions.
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: I think you need more description here (you had a comment in your PR about adding javadocs, so maybe this is intentional).  This should include a detailed description of exactly how it works and any restrictions it has.  For example: 

* It is provided a generator that produces a cursor
* The cursor is iterated over until it is either exhausted or until scan limit properties are reached
   * if scan limit properties are reached, it re-starts the cursor in the context of a new transaction
* Once a cursor is exhausted, the generator is asked for the next cursors and the process is repeated
* It is responsible for all `FDBRecordContext` management, so all reads it does are in the scope of transactions that it controls.
* The cursor has no visibility into the `ScanProperties` of cursors returned by the generator and, therefore, will not be involved in enforcing any limits that may be, individually, applied to the `ScanProperties` that may be applied to the cursor returned by the generator.  For example, if the generator returns a cursor that specified, say, a record scan limit of 10 record, the `AutoContinuingCursor` will scan all data until it is exhausted, 10 records at a time.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        if (currentCursor == null) {
+            openContextAndGenerateCursor(null);
+        }
+        return currentCursor.onNext().thenCompose(recordCursorResult -> {
+            if (recordCursorResult.noNextButHasContinuation()) {
+                currentContext.close();
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Maybe to be on the safe side, you should move the logic to close the context into `openContextAndGeneratorCursor` like:

    if (currenContext != null) {
        currentContext.close();
    }

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/ChainedCursor.java [hunk 4]
```diff
@@ -51,8 +51,8 @@
  *             } else {
  *                 return Optional.empty();
  *             }
- *             return Optional.of(1);
  *         }
+ *         return Optional.of(1);
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/ChainedCursor.java]: Thanks :-)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java (hunk 1 above)
Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java (hunk 2 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Yes, I had a weird mental hiccup with what the cursor was supposed to do :-) 

However, it is *possible* that the user wanted to provide a cursor generator that chained together different cursors.  I.e., use cursor A until exhausted, then use cursor B until exhausted. So just clarifying that once a returned cursor declares its source is exhausted, then the `AutoContinuingCursor` is exhausted.

Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Yes. See `If the <i>underlying cursor</i> is exhausted, the {@link AutoContinuingCursor} is also exhausted.`

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java [hunk 5]
```diff
@@ -200,6 +200,14 @@ public int hashCode() {
         }
     }
 
+    /**
+     * Returns true if the cursor has reached its end but a continuation is still available (i.e. the source is not yet exhausted).
+     * @return {@code true} if this result does not include a next value but include a continuation and {@code false} otherwise
+     */
+    public boolean noNextButHasContinuation() {
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java]: This method seems very useful but somehow the name seems a bit confusing. How do you feel about `isStoppedWithContinuation()` or `hasStoppedBeforeEnd()` or something along those lines? This is a key part of our external API so I thin it's worth getting it right (so it can be stable).

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java [hunk 6]
```diff
@@ -200,6 +200,14 @@ public int hashCode() {
         }
     }
 
+    /**
+     * Returns true if the cursor has reached its end but a continuation is still available (i.e. the source is not yet exhausted).
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java]: `true` -> `{@code true}`.

Perhaps it's clearer to change "a continuation is still available" to "the continuation is not an end continuation" because a continuation is always "available" (it might just be `END`)?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+
+/**
+ * A cursor that can iterate over a cursor across transactions.
+ *
+ * It is provided a generator that produces a cursor (referred to as <i>underlying cursor</i>). The <i>underlying
+ * cursor</i> is iterated over until it is either exhausted or until scan limit properties are reached:
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: "scan limit properties are reached" -> "one of the scan limits is reached", for consistency with other docs.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+ * <p>
+ * {@link AutoContinuingCursor} is responsible for all {@link FDBRecordContext} management, so all reads it does are in
+ * the scope of transactions that it controls.
+ * </p>
+ *
+ * <p>
+ * The {@link AutoContinuingCursor} has no visibility into the {@link com.apple.foundationdb.record.ScanProperties} of
+ * the <i>underlying cursor</i> and, therefore, will not be involved in enforcing any limits that may be individually
+ * applied to the <i>underlying cursor</i>. For example, if the generator returns an <i>underlying cursor</i> that
+ * specified, say, a record scan limit of 10 record, the {@link AutoContinuingCursor} will scan all data until it is
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: "10 record" -> "10 records"

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        if (currentCursor == null) {
+            openContextAndGenerateCursor(null);
+        }
+        return currentCursor.onNext().thenCompose(recordCursorResult -> {
+            if (recordCursorResult.noNextButHasContinuation()) {
+                openContextAndGenerateCursor(lastResult.getContinuation().toBytes());
+                return currentCursor.onNext().thenApply(finalRecordCursorResult -> {
+                    if (finalRecordCursorResult.noNextButHasContinuation()) {
+                        throw new RecordCoreException("failed to continue the cursor in a new context");
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: I'm not 100% sure of the semantics that we want here but this could totally happen in practice and it seems okay to me.

For example, suppose that we have a very selective `FilterCursor` around a `KeyValueCursor` and a relatively low scan limit. In that case, it will be totally normal to have a cursor that we just opened return a `RecordCursorResult` with a continuation and a `NoNextReason` of `SCAN_LIMIT_REACHED`, which seems like it would cause an exception here. Maybe that's what we actually want here, but if it is we should document it carefully.

Alternatively, it seems reasonable to use `AsyncUtils.whileTrue()` to keep generating new cursors (perhaps in the same context?) until we get some kind of result from `currentCursor.onNext()`.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/cursors/AutoContinuingCursorTest.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+            );
+
+            List<Integer> returnedList = new ArrayList<>();
+            while (cursor.hasNext()) {
+                returnedList.add(cursor.next());
+            }
+
+            assertEquals(list, returnedList);
+        }
+    }
```
Comment (nschiefer) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/cursors/AutoContinuingCursorTest.java]: Regardless of the semantics that we decide on, I'd like to see a test that makes assertions about the semantics of a  cursor that usually gets "stuck" before returning results (like I described above). Maybe that's an `assertThrows` about the exception, but it's worth checking for regressions.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java (hunk 9 above)
Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: My concern was it might be an infinite loop if the generated cursor always return no next given a new cursor. But maybe it is expected.

@scgray What do you think?

Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: @nblintao I agree with @nschiefer - we should continue to loop with new contexts as long as the cursor is making progress, whether or not that progress is returning a value.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java (hunk 5 above)
Comment (scgray) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursorResult.java]: Yeah, I like `hasStoppedBeforeEnd()`

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+ * the <i>underlying cursor</i> and, therefore, will not be involved in enforcing any limits that may be individually
+ * applied to the <i>underlying cursor</i>. For example, if the generator returns an <i>underlying cursor</i> that
+ * specified, say, a record scan limit of 10 records, the {@link AutoContinuingCursor} will scan all data until it is
+ * exhausted, at most 10 records at a transaction.
+ * </p>
+ *
+ * @param <T> the type of elements returned by this cursor
+ */
+@API(API.Status.EXPERIMENTAL)
+public class AutoContinuingCursor<T> implements BaseCursor<T> {
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Since the `AutoContinuingCursor` has a child and doesn't enforce the scan limits, I don't think it should be a `BaseCursor`. It should probably just implement `RecordCursor<T>` instead.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return runner.getExecutor();
+    }
+
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
```
Comment (nschiefer) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Because `AutoContinuingCursor` has a child cursor, it should probably pass the visitor down to its (current) child. I think this method should look something like:
```java
if (visitor.visitEnter(this)) {
    currentCursor.accept(visitor);
}
return visitor.visitLeave(this);
```

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java (hunk 11 above)
Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java (hunk 12 above)
Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Done.

**Comment by nblintao** [comment]
Hold on, I have a wrong base.

**Comment by nblintao** [comment]
OK to go.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java (hunk 12 above)
Comment (nschiefer) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Ack, `currentCursor` is nullable, right? I guess we need to check for that, too :(

Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java]: Updated in the same commit.

**Comment by nblintao** [comment]
@nschiefer Yes, please. Thanks!
